nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+
+- [Nexus RPCs - short recap](#nexus-rpcs---short-recap)
+- [Operation identifiers](#operation-identifiers)
+  - [Caller generated identifiers](#caller-generated-identifiers)
+    - [Handler addressability](#handler-addressability)
+      - [Mapping from client to handler identifiers](#mapping-from-client-to-handler-identifiers)
+  - [Handler generated identifiers](#handler-generated-identifiers)
+    - [Cancel before started](#cancel-before-started)
+  - [Comparing the approaches](#comparing-the-approaches)
+    - [Handler Id qualification](#handler-id-qualification)
+    - [Ids as inputs](#ids-as-inputs)
+    - [Caller friendliness](#caller-friendliness)
+    - [Data sensitivity](#data-sensitivity)
+- [Proposal](#proposal)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Nexus RPCs - short recap
+
+At the fundamental level, the Nexus protocol is a synchronous RPC protocol. Asynchronous operations are modelled on top
+of a set of pre-defined RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. These asynchronous operations can be cancelled and checked for their outcomes and current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation's when it
+is ready.
+
+## Operation identifiers
+
+Since an asynchronous operation is modelled on top of synchronous RPCs, it must be addressable (identifiable) by both
+the caller and the handler. There are two approaches for generating identifiers, caller side and handler side.
+
+### Caller generated identifiers
+
+With this approach, the caller provides an Id to each of the pre-defined RPCs.
+
+The handler _should_ qualify an identifier when dealing with untrusted sources and multi-tenant applications.
+
+The following set of RPCs would be needed to model asynchronous operations using client generated Ids:
+
+- `Start(operation, id, input, callbackURL)` - handler responds inline or indicates that an async operation was started
+- `Cancel(operation, id)` - cancel operation by Id
+- `GetResult(operation, id)` - get operation result by Id
+- `Describe(operation, id)` - get operation status by Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+Example invocations:
+
+```js
+Start(
+  'charge',
+  transactionId,
+  { customerId: 'someCustomerId', amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', transactionId);
+```
+
+#### Handler addressability
+
+In some cases, the handler might require more than just a single identifier for a given operation, e.g. when an
+operation is backed by a workflow update, which requires both a workflow Id and an update Id. In order to avoid
+maintaining a handler side persistent mapping between caller and handler identifiers, the handler may choose to "leak"
```

Comment:
I don't see the benefit of this "middle-of-the-ground". What the server returns should be opaque to the client, so that we can easily change things without breaking clients, or support different backend implementations.  Leaking implementation details is a bad idea. Forcing the client to do extra work is even worse. And you have to maintain the mappings in the handler anyway...

Reply:
> Forcing the client to do extra work is even worse

Forcing the client to store an identifier that can only be generated server side is cording the client to do extra work

Reply:
Point taken. But I think we can help stateless clients to persist and manage these aliases with the right service, and if we get the right client abstraction, it won't be too painful...

Reply:
That's what the persistent mapping is for, but the handler would decide when to use it, not the caller.

Reply:
Can I resolve this thread?